# Moffitt_Internship_2023
![Figure Of RawCount and H5 Protocol](https://github.com/Gabrielle-Nobles/Moffitt_Internship_2023/assets/97853225/d0e210b2-0dc6-4b6e-b07b-b05fb3b24cb2)

## Introduction 
The utilization of single-cell RNA sequencing (scRNA-seq) has emerged as a robust and potent approach to examining the complexities of cellular heterogeneity and genetic expression with distinctive resolution. However, analyzing scRNAseq data can be complex and requires a standardized approach. This protocol outlines a comprehensive workflow for analyzing scRNAseq data using R and the Seurat package, along with other tools like SeuratDisk, pathwork, dplyr, Single R, and Celldex. The protocol covers essential steps such as quality control, normalization, data scaling, dimensionality reduction, clustering, and automated cell annotation. Following this protocol allows bioinformaticians and researchers to effectively analyze scRNAseq data, identify distinct cell populations, and gain valuable insights into cellular dynamics and functions. The output files generated by this protocol, including metadata, H5 Seurat files, cell subpopulation metadata, and ISCVA-compliant files, facilitate downstream analyses and enable integration with other analysis and visualization tools. This protocol provides a standardized and reproducible framework for scRNAseq analysis. 
## Installation 

### R dependencies 

## Protocol 1: Raw Counts 
### Input Files 
To achieve accurate results in single-cell RNA sequencing (scRNAseq) analysis, it's crucial to have high-quality and compatible input data. The input file should be a raw count matrix of single cells, with each row representing a gene and each column representing a specific cell. The numbers within the matrix indicate the raw expression counts or read counts of genes in each cell. Ensure that the input file is in a compatible format, such as a Comma-Separated Values (CSV) file, tab-delimited text file or a Tab-Seperated Values (TSV) file, as these formats are easily readable and processed by R and the Seurat package. If you have used the CellRanger pipeline to process your scRNAseq data, you should have three input files: matrix (matrix.mtx), barcode (barcode.tsv), and features (features.tsv).
## Single Cell RNA-seq Analysis 
1. Quality Control
- Perform quality control (QC) steps to filter out low-quality cells and genes. QC calculates the percentages of mitochondrial genes and ribosomal protein genes for each cell, and then filters out cells that have high mitochondrial gene content and low detected features in the RNA assay.  
3. Normalization 
- After completing the QC steps, it's important to normalize the count matrix to adjust for differences in library sizes and sequencing depth. This is a crucial step in analyzing scRNAseq data because it accounts for variations in library sizes and sequencing depth between individual cells. Normalization ensures that expression values can be compared accurately and eliminates technical biases. 
- Our method employs the "LogNormalize" parameter, which performs a global-scaling normalization. It divides each gene's expression by the total expression of that gene across all cells, multiplies the data by a default scale factor of 10,000, and log-transforms it. This normalization method is valuable for preserving the relative differences between cells while normalizing gene expression across them.
5. Scaling Data 
6. Prinicpal Component Analysis (PCA)
7. Nearest Neighbor 
8. SNN clustering 
9. Automated Cell Annotation using SingleR and Celldex

## Protocol 2: H5 Seurat Protocol 
### Required Files 



